File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/CustomPinTypeRepository.kt
package com.rork.vinetrack.data

import com.rork.vinetrack.data.auth.SessionStore
import com.rork.vinetrack.data.model.CustomPinCreateParams
import com.rork.vinetrack.data.model.CustomPinType
import com.rork.vinetrack.data.model.CustomPinTypeCreateParams
import com.rork.vinetrack.data.model.ManualIssue
import io.ktor.client.call.body
import io.ktor.client.request.headers
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Server-authoritative access to the sql/170 unified pin composer RPCs:
 * the vineyard-shared custom pin type catalogue and the simplified Custom
 * pin create. Mirrors iOS `CustomPinTypeRepository`.
 *
 * Idempotency: both creates are keyed by client-generated ids, so an offline
 * replay returns the existing canonical row instead of duplicating; a
 * duplicate active type NAME (trimmed, case-insensitive) converges on the
 * existing shared entry server-side.
 */
class CustomPinTypeRepository(private val session: SessionStore) {

    @Serializable
    private data class ListArgs(
        @SerialName("p_vineyard_id") val vineyardId: String,
        @SerialName("p_include_inactive") val includeInactive: Boolean = false,
    )

    @Serializable
    private data class SetActiveArgs(
        @SerialName("p_id") val id: String,
        @SerialName("p_is_active") val isActive: Boolean,
    )

    suspend fun createType(params: CustomPinTypeCreateParams): CustomPinType =
        rpc("create_vineyard_custom_pin_type", params)

    suspend fun listTypes(vineyardId: String, includeInactive: Boolean = false): List<CustomPinType> =
        rpc("list_vineyard_custom_pin_types", ListArgs(vineyardId, includeInactive))

    suspend fun setTypeActive(id: String, isActive: Boolean): CustomPinType =
        rpc("set_vineyard_custom_pin_type_active", SetActiveArgs(id, isActive))

    /** Simplified Custom-tab save — returns the canonical manual-issue JSON. */
    suspend fun createCustomPin(params: CustomPinCreateParams): ManualIssue =
        rpc("create_custom_pin", params)

    private suspend inline fun <reified Body, reified Result> rpc(name: String, body: Body): Result =
        withContext(Dispatchers.IO) {
            if (!SupabaseClient.isConfigured) throw BackendError.NotConfigured
            val token = session.accessToken ?: throw BackendError.Unauthorized
            val response = SupabaseClient.http.post(SupabaseClient.rpcUrl(name)) {
                headers {
                    append("apikey", SupabaseClient.anonKey)
                    append("Authorization", "Bearer $token")
                }
                contentType(ContentType.Application.Json)
                setBody(body)
            }
            when {
                response.status.isSuccess() -> response.body<Result>()
                response.status.value == 401 || response.status.value == 403 -> throw BackendError.Unauthorized
                else -> throw BackendError.Server(response.status.value, response.bodyAsText())
            }
        }
}

File: ios/VineTrack/Backend/Repositories/CustomPinTypeRepository.swift
import Foundation
import Supabase

/// Server-authoritative access to the sql/170 unified pin composer RPCs:
/// the vineyard-shared custom pin type catalogue, the simplified Custom pin
/// create, and the generic row-segment persistence used by Repair/Growth
/// pins saved with a ROW location. Mirrors `CustomPinTypeRepository.kt`.
final class CustomPinTypeRepository {
    private let provider: SupabaseClientProvider

    init(provider: SupabaseClientProvider = .shared) {
        self.provider = provider
    }

    /// Idempotent create keyed by the client-generated id; a duplicate active
    /// name (trimmed, case-insensitive) returns the existing shared entry.
    func createType(_ params: CustomPinTypeCreateParams) async throws -> CustomPinTypeRecord {
        guard provider.isConfigured else { throw BackendRepositoryError.missingSupabaseConfiguration }
        return try await provider.client
            .rpc("create_vineyard_custom_pin_type", params: params)
            .execute()
            .value
    }

    /// Active items by default (what the composer offers).
    func listTypes(vineyardId: UUID, includeInactive: Bool = false) async throws -> [CustomPinTypeRecord] {
        guard provider.isConfigured else { throw BackendRepositoryError.missingSupabaseConfiguration }
        return try await provider.client
            .rpc(
                "list_vineyard_custom_pin_types",
                params: ListTypesParams(vineyardId: vineyardId, includeInactive: includeInactive)
            )
            .execute()
            .value
    }

    /// Deactivating hides an item from new selection; historical pins keep
    /// their meaning.
    func setTypeActive(id: UUID, isActive: Bool) async throws -> CustomPinTypeRecord {
        guard provider.isConfigured else { throw BackendRepositoryError.missingSupabaseConfiguration }
        return try await provider.client
            .rpc(
                "set_vineyard_custom_pin_type_active",
                params: SetActiveParams(id: id, isActive: isActive)
            )
            .execute()
            .value
    }

    /// Simplified Custom-tab save — returns the canonical manual-issue JSON.
    func createCustomPin(_ params: CustomPinCreateParams) async throws -> ManualIssueRecord {
        guard provider.isConfigured else { throw BackendRepositoryError.missingSupabaseConfiguration }
        return try await provider.client
            .rpc("create_custom_pin", params: params)
            .execute()
            .value
    }

    /// Persist the structured ROW selection for a Repair/Growth pin created
    /// through the existing direct pin write path. Raises PIN_NOT_FOUND while
    /// the parent pin hasn't reached the server yet — callers treat that as
    /// retryable.
    func setPinRowSegments(pinId: UUID, segments: [ManualIssueSegment]) async throws -> [ManualIssueSegment] {
        guard provider.isConfigured else { throw BackendRepositoryError.missingSupabaseConfiguration }
        return try await provider.client
            .rpc(
                "set_pin_row_segments",
                params: SegmentsParams(pinId: pinId, segments: segments)
            )
            .execute()
            .value
    }
}

nonisolated private struct ListTypesParams: Encodable, Sendable {
    let vineyardId: UUID
    let includeInactive: Bool

    enum CodingKeys: String, CodingKey {
        case vineyardId = "p_vineyard_id"
        case includeInactive = "p_include_inactive"
    }
}

nonisolated private struct SetActiveParams: Encodable, Sendable {
    let id: UUID
    let isActive: Bool

    enum CodingKeys: String, CodingKey {
        case id = "p_id"
        case isActive = "p_is_active"
    }
}

nonisolated private struct SegmentsParams: Encodable, Sendable {
    let pinId: UUID
    let segments: [ManualIssueSegment]

    enum CodingKeys: String, CodingKey {
        case pinId = "p_pin_id"
        case segments = "p_segments"
    }
}
